middleware/2fa.js

```diff
@@ -0,0 +1,287 @@
+    ctx.body = {
+        success: true,
+        publicKey
+    };
+};
+// https://near-contract-helper-2fa.onrender.com/
+// http post http://localhost:3000/2fa/init accountId=mattlock method:='{"kind":"2fa-email","detail":"matt@near.org"}'
+// http post https://near-contract-helper-2fa.onrender.com/2fa/init accountId=mattlock method:='{"kind":"2fa-email","detail":"matt@near.org"}'
+// Call ONCE to enable 2fa on this account. Adds a twoFactorMethod (passed in body) where kind should start with '2fa-'
+// This WILL send the initial code to the method specified ['2fa-email', '2fa-phone']
+const initCode = async (ctx) => {
+    const { accountId, method, testing, code_hash } = ctx.request.body;
+    const [account] = await models.Account.findOrCreate({ where: { accountId } });
+    if (!account) {
+        ctx.throw(401, 'account should be created');
+        return;
+    }
+    if (!method || !method.kind || !method.detail) {
+        ctx.throw(401, 'method arguments invalid');
+        return;
+    }
+    const { kind, detail } = method;
+    // validate method kind
+    if (!twoFactorMethods.includes(kind)) {
+        ctx.throw(401, 'invalid 2fa method ' + kind);
+        return;
+    }
+    const hasContractDeployed = await isContractDeployed(accountId);
+    // check recover methods
+    let [twoFactorMethod] = await account.getRecoveryMethods({ where: {
+        kind: {
+            [Op.startsWith]: '2fa-'
+        },
+    }});
+    // check if multisig contract is already deployed
+    if (hasContractDeployed || !!code_hash) {
+        // check to see if they already have at least 1 twoFactorMethod
+        if (twoFactorMethod) {
+            ctx.throw(401, 'account with multisig contract already has 2fa method');
+        } else {
+            // unlikely
+            twoFactorMethod = await account.createRecoveryMethod({ kind, detail, requestId: -1 });
+        }
+    } else {
+        // as long as the multisig is not deployed, can keep updating (or create new, 2fa method)
+        if (!twoFactorMethod) {
+            twoFactorMethod = await account.createRecoveryMethod({ kind, detail, requestId: -1 });
+        } else {
+            await twoFactorMethod.update({
+                kind: method.kind,
+                detail: method.detail,
+                requestId: -1
+            });
+        }
+    }
+    // check if 2fa method matches existing recovery method
+    const [recoveryMethod] = await account.getRecoveryMethods({ where: {
+        kind: kind.split('2fa-')[1],
+        detail,
+    }});
+    if (!recoveryMethod) {
+        // client waits to deploy contract until code is verified
+        const securityCode = await sendCode(method, twoFactorMethod);
+        const body = {
+            success: true,
+            message: '2fa initialized and code sent to verify method',
+        };
+        if (testing) {
+            body.securityCode = securityCode;
+        }
+        ctx.body = body;
+    } else {
```

Comment:
nit: I think makes sense to use early return for this to avoid unnecessary nesting https://www.itamarweiss.com/personal/2018/02/28/return-early-pattern.html